File: lib/src/test/java/it/cosenonjaviste/daggermock/dependency/ComponentDependencyTest.java
package it.cosenonjaviste.daggermock.dependency;

import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;

import it.cosenonjaviste.daggermock.DaggerMockRule;

import static org.assertj.core.api.Assertions.assertThat;

public class ComponentDependencyTest {

    @Rule
    public final DaggerMockRule<MyComponent> mockitoRule = new DaggerMockRule<>(MyComponent.class, new MyModule())
            .addComponentDependency(MyComponent2.class, new MyModule2())
            .set(new DaggerMockRule.ComponentSetter<MyComponent>() {
                @Override public void setComponent(MyComponent component) {
                    mainService = component.mainService();
                }
            });

    @Mock
    MyService myService;
    @Mock
    MyService2 myService2;

    private MainService mainService;

    @Test
    public void testComponentDependencyModulesCanBeOverriden() {
        assertThat(mainService).isNotNull();
        assertThat(mainService.getMyService()).isSameAs(myService);
        assertThat(mainService.getMyService2()).isSameAs(myService2);
    }
}

File: daggermockTests/src/test/java/it/cosenonjaviste/daggermock/builder/ComponentDependencyTest.kt
package it.cosenonjaviste.daggermock.builder

import com.nhaarman.mockitokotlin2.mock
import it.cosenonjaviste.daggermock.DaggerMockRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test

class ComponentDependencyTest {

    @get:Rule
    val rule = DaggerMockRule<MyComponent>(MyComponent::class.java, MyModule())
            .addComponentDependency(MyComponent2::class.java, MyModule2())
            .set { component -> mainService = component.mainService() }

    private val myService: MyService = mock()
    private val myService2: MyService2 = mock()

    private lateinit var mainService: MainService

    @Test
    fun testComponentDependencyModulesCanBeOverriden() {
        assertThat(mainService.myService).isSameAs(myService)
        assertThat(mainService.myService2).isSameAs(myService2)
    }
}
